### ISSUE TITLE
Kafka Commit is not working

### ISSUE THREAD

**Issue by alok87** [Shopify/sarama#1894]
##### Versions

| Sarama | Kafka | Go |
|--------|-------|----|
v1.27.2 | 2.6.0 | 1.15.7|
##### Configuration

What configuration values are you using for Sarama and Kafka?
``` yaml
Assignor:   "range",
Oldest:     true,
Log:        false,
AutoCommit: false,
```

##### Problem Description

Our Consumer group takes a lot of time to process based on the size of the message batch it handles. 

**Problem**: What we have seen is when the group is working on consuming claim for 100+ topics partition concurrently and the batch size is huge then the manual commit does not work sometimes. Test 1 as described here is not working for me. 

https://github.com/Shopify/sarama/issues/1570#issuecomment-789332886

**What does not work**
1. Big batch (msgBuf)
2. Auto commit = false

**What works**
1. Smaller batch (msgBuf)
2. Auto commit = true (from false)

Below is the `markOffset` code for the batch of messages. We run this when all msgs in the msgBuf are consumed, we mark only the last message and the commit.

```go
        if len(msgBuf) > 0 {
		lastMessage := msgBuf[len(msgBuf)-1]
		b.session.MarkOffset(
			lastMessage.Topic,
			lastMessage.Partition,
			lastMessage.Offset+1,
			"",
		)

		if b.autoCommit == false {
			b.session.Commit()
			klog.V(2).Infof("%s, Committed (autoCommit=false)", lastMessage.Topic)
		}

		b.lastCommittedOffset = lastMessage.Offset

	} else {
		klog.Warningf("%s, markOffset not possible for empty batch", b.topic)
	}
```

The `b.session.Commit()` happens without an issue but the offset never gets committed. Commit is a synchronous call but returns no errors. So we are not really sure at present why the commit does not take effect. Both markoffset and commit calls happen without any problem. The next restart starts processing from the old offset position and it gets stuck in this reprocessing loop.

Could it be because of dead session `b.session` but the call throws no error.

**What do you think could be the cause of this issue?**

**Comment by alok87** [comment]
## Tests 
Most likely have figured out the issue. 
```
Test1: maxSizeOfBatch: 3, autoCommit: true ==> works ✅ (it is = 3*10k messages)
Test2: maxSizeOfBatch: 100, autoCommit: true ==> works ✅ (it is = 100*10k messages)
Test3: maxSizeOfBatch: 100, autoCommit: false ==> works ✅ (it is = 100*10k messages)
Test4: maxSizeOfBatch: 1000, autoCommit: true ==> ❌  (it is = 1000*10k messages)
```

Test4 issue got reproduced with auto commit true and large batch size. 

``` test 4 result
I0303 04:16:16.658142       1 load_processor.go:151] customers, offset: 3367, marking
I0303 04:16:16.658147       1 load_processor.go:158] customers, offset: 3367, marked and commited
I0303 04:16:16.658151       1 load_processor.go:679] customers, batchId:1, startOffset:2367, endOffset:3366: Processed
```
We saw the commit finally started showing as 3366 only 10 mins after at 04:26:02

## Conclusion
So when the batch size is really large, we see the marking every message takes a lot of time. It reflected the But the weird thing here is we are only **marking** only the last offset in the batch and have kept auto commit true. 

## Question
How is the batch size here creating issue?

**Comment by alok87** [comment]
Closed as it is not a sarama issue. 
| Why is the commit for one offset taking time 10mins when autoCommit=true is the question.
this ^ is still unanswered at present but nothing to do with sarama.

**Comment by alok87** [comment]
Tracing what happens in sarama's session.Commit(), I saw the following errors
`offset_manager.go:340  kafka server: The provided member is not known in the current generation. `
Then I found this Shopify/sarama#1310 which says you get the generation error when you are trying to commit from a closed session.
This is the reason big batches were not working out.
Confirmed and reproduced this issue. Increased the **MaxProcessingTime** and lived happily ever after.
